control-plane/agents/core/src/volume/specs.rs

```diff
@@ -464,9 +436,416 @@ impl ResourceSpecsLocked {
+                        break;
+                    } else {
+                        // this replica is no longer valid
+                        // todo: do it now, or let the reconcile do it?
+                        spec_replica.owners.disowned_by_volume();
+                    }
+                }
+            }
+        }
+        let mut nexus_replicas = vec![];
+        // now reduce the replicas
+        // one replica per node, with the right share protocol
+        for (spec, status) in spec_status_pair.iter() {
+            if nexus_replicas.len() > vol_spec.num_replicas as usize {
+                // we have enough replicas as per the volume spec
+                break;
+            }
+            let (share, unshare) = {
+                let spec = spec.lock().await;
+                let local = &status.node == target_node;
+                (
+                    local && (spec.share.shared() | status.share.shared()),
+                    !local && (!spec.share.shared() | !status.share.shared()),
+                )
+            };
+            if share {
+                // unshare the replica
+                if let Ok(uri) = self.unshare_replica(registry, &status.into()).await {
+                    nexus_replicas.push(ChildUri::from(uri));
+                }
+            } else if unshare {
+                // share the replica
+                if let Ok(uri) = self.share_replica(registry, &status.into()).await {
+                    nexus_replicas.push(ChildUri::from(uri));
+                }
+            } else {
+                nexus_replicas.push(ChildUri::from(&status.uri));
+            }
+        }
+
+        // Create the nexus on the request.node
+        self.create_nexus(
+            registry,
+            &CreateNexus {
+                node: target_node.clone(),
+                uuid: NexusId::new(),
+                size: vol_spec.size,
+                children: nexus_replicas,
+                managed: true,
+                owner: Some(vol_spec.uuid.clone()),
+            },
+        )
+        .await
+    }
+
     /// Delete volume by its `id`
     async fn del_volume(&self, id: &VolumeId) {
         let mut specs = self.write().await;
         specs.volumes.remove(id);
     }
 }
+
+fn get_volume_nexus(volume_status: &Volume) -> Result<Nexus, SvcError> {
+    match volume_status.children.len() {
+        0 => Err(SvcError::VolumeNotPublished {
+            vol_id: volume_status.uuid.to_string(),
+        }),
+        1 => Ok(volume_status.children[0].clone()),
+        _ => Err(SvcError::NotReady {
+            kind: ResourceKind::Volume,
+            id: volume_status.uuid.to_string(),
+        }),
+    }
+}
+
+async fn get_volume_target_node(
+    registry: &Registry,
+    status: &Volume,
+    request: &PublishVolume,
+) -> Result<NodeId, SvcError> {
+    // If the nexus is already published, then it should first be unpublished
```

Comment:
This comment mentions unpublishing, but we don't do that here. I guess that's a decision for the caller. Does the comment need updating?

Reply:
Yep, updated to: "// We can't configure a new target_node if the volume is currently published"

Reply:
:+1: